### DISCUSSION TITLE
allow implicit target type deduction

### DISCUSSION THREAD

**Discussion by kasperk81** [dotnet/csharplang#8461] [General]
`foreach (var n in [1,2,3])` rightfully throws `error CS9176: There is no target type for the collection expression.` and expects `(int[])` cast.

compiler should at least be able to deduce int[] implicitly when type is specified `foreach (int n in [1,2,3])`?

**Comment by HaloFour** [comment]
See: #7913

For a `for` loop I would hope that the compiler can be smart enough to elide allocations, so not even `int[]`.
